### DISCUSSION TITLE
extracting the motif from stumpy.motifs

### DISCUSSION THREAD

**Discussion by adityabhandwalkar** [TDAmeritrade/stumpy#795] [Help: Coding & Implementations]
Hello everyone,

I am trying to implement the `.motifs` from stumpy to get top k motifs using the following code

```
k  =  input('Enter the number of top k motifs to be discovered?\n')
              mp_dist, mp_idx = stumpy.motifs(data, mp[:,0],max_motifs= int(k),max_distance=None,max_matches=10)             

      # plot the top k motifs out of mp_idx
              for motif in mp_idx:
                    plt.plot(data[motif:motif + self.window_size], lw = 2, label ='top k-motifs' )
              plt.show()
```

While doing so it throws an error with the following message
`TypeError: only integer scalar arrays can be converted to a scalar index`
both` mp_idx` and `data` is of type `ndarray`
Can anyone tell me what is  wrong with this code ?

thanks

**Comment by seanlaw** [comment]
@adityabhandwalkar Can you please provide a minimum-reproducible-example including the exact data and code that causes the problem? Otherwise, it is impossible to know what the issue is. Based on the limited information that you've provided, this doesn't sound like a STUMPY problem but likely a problem with your inputs.

**Comment by adityabhandwalkar** [comment]
```
_data = [-0.00639439  0.00245094  0.00985026 ... -0.01012683  0.00349879
  0.01501203]_
```
data type is of numpy.ndarry 
code:

 ```
def profile(self,data):
              ''' Calculating matrix profile with window_size predefined '''
              
              mp = stumpy.stump(data, m= self.window_size)
              
         # discord
              mp_discord= np.argsort(mp[:, 0])[-1]
       # similar pattern
              mp_motif = np.argsort(mp[:, 0])[0]
              nearest_neighbor_idx = mp[mp_motif, 1]
             
        # finding top motifs with the matrix profile
             
              k  =  input('Enter the number of top k motifs to be discovered?\n') # k =5
              mp_dist, mp_idx =  (data, mp[:,0],max_motifs= int(k),max_distance=None,max_matches=10)             

      # plot the top k motifs out of mp_idx
              for motif in range(mp_idx):
                    plt.plot(data[motif:motif + self.window_size], lw = 2, label ='top k-motifs' )
              plt.show()
```

**Reply by adityabhandwalkar** [reply]
window size is 2048

**Reply by seanlaw** [reply]
@adityabhandwalkar The code that you provided has some basic problems that make it very hard to assist you:

1. It looks like the input array that you've provided was copy and pasted and therefore has been truncated and only shows 6 values above. Next time, please save the data as a CSV file and upload it here (by dragging and dropping the file into a comment box)
2. For a minimum-reproducible-example, please remove/omit ALL unnecessary code.  Currently, your code is wrapped in a function and the (Python) indentation is bad. This makes it very difficult for others to reproduce your error and wastes a lot of time to correct. When I copy and paste your code, the Python interpreter complains as you also have missing, undeclared variables. The best minimum-reproducible-examples allow others to simply copy, paste, and execute to reproduce your error directly. Once you think you've produced a minimum-reproducible-example, fire up a new Python environment,  and try to execute your code there.
3. You are missing the actual call to `stumpy.motifs`
4. Please provide the full Python error traceback

Having said all that, here is a minimum reproducible example based on your code (note that I was forced to waste a lot of time to fix the formatting before I could get to your actual error):

```
import stumpy
import numpy as np

data = np.random.rand(10000)
m = 2048

mp = stumpy.stump(data, m)             
mp_discord= np.argsort(mp[:, 0])[-1]
mp_motif = np.argsort(mp[:, 0])[0]
nearest_neighbor_idx = mp[mp_motif, 1]

k = 5
mp_dist, mp_idx = stumpy.motifs(data, mp[:,0], max_motifs=int(k), max_distance=None, max_matches=10)             

for motif in range(mp_idx):
    plt.plot(data[motif:motif + m], lw = 2, label ='top k-motifs' )
```

This produces the following Python tracback:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 15
     12 k = 5
     13 mp_dist, mp_idx = stumpy.motifs(data, mp[:,0], max_motifs=int(k), max_distance=None, max_matches=10)             
---> 15 for motif in range(mp_idx):
     16     plt.plot(data[motif:motif + m], lw = 2, label ='top k-motifs' )

TypeError: only integer scalar arrays can be converted to a scalar index
```

Notice that it tells you WHERE the error is coming from. Namely, 

```
---> 15 for motif in range(mp_idx):
```

This is NOT a problem with STUMPY and, instead, is a problem with the input to `range`, which is `mp_idx`, If you `print(mp_idx)`, you'll see:

```
array([[4485, 7250, 3895, 5489, 6302, 2922, 2313,   49, 1560]])
```

In other words, `mp_idx` is a 2D array and Python's  `range` function doesn't know what do with this as it is expecting an `integer scalar array` (i.e., a 1D array). There are multiple different ways to overcome this but the most obvious way is to index into the first row of the 2D array:

```
for row in mp_idx:
    for motif in row:
        plt.plot(data[motif:motif + m], lw = 2, label ='top k-motifs' )
```

@adityabhandwalkar I want to reiterate again that this is NOT a STUMPY problem. I recommend spending some time learning about the Python `range` function and reviewing some Python basics. This will help you in your work and it will make it much easier for others to assist you as we have limited time and can only really help with STUMPY problems and not basic Python problems. For basic Python problems (unrelated to STUMPY), I recommend posting your questions to StackOverflow. I hope this helps.
